.github/workflows/testing-and-deployment.yml

```diff
@@ -217,7 +231,14 @@ jobs:
 
       - name: Plotting Testing (uses GL)
         continue-on-error: ${{ matrix.numpy-version == 'pre' }}
-        run: xvfb-run -a python -m pytest --cov=pyvista --cov-append --cov-report=xml --fail_extra_image_cache -v --ignore=tests/core --ignore=tests/examples
+        run: xvfb-run -a python -m pytest --cov=pyvista --cov-append --cov-report=xml --fail_extra_image_cache -v --ignore=tests/core --ignore=tests/examples --generated_image_dir _debug_images
+
+      - name: Upload Generated Images
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: ${{ github.job }}
```

Comment:
```suggestion
          name: ${{ github.job }}-${{ matrix.python-version }}-${{ matrix.vtk-version }}-${{ matrix.numpy-version }}
```

Reply:
This suggestion is better than what was there before. I'm happy with this (and the other) suggestions.

Ideally there would be a single magic variable that returns the full test name instead of needing to concatenate the relevant vars, e.g. if the numpy pre-release tests are removed in the future, it would be preferred not to also need to remove the refs to `matrix.numpy-version`. Not sure if this is possible though.

Reply:
> Ideally there would be a single magic variable that returns the full test name instead of needing to concatenate the relevant vars, e.g. if the numpy pre-release tests are removed in the future, it would be preferred not to also need to remove the refs to `matrix.numpy-version`. Not sure if this is possible though.

This is what I was going for in my comments, thanks for understanding my terse statements.  I fear that this method will yield a situation where it becomes a maintenance burden for CI.  I'm also looking into this, otherwise, this type of suggestion looks like a good path forward.

Reply:
That's too bad there isn't a simple way to do this. Maybe it's possible to simply use `matrix` generically?
``` diff
+ ${{ github.job }}-${{ matrix }}
- ${{ github.job }}-${{ matrix.python-version }}-${{ matrix.vtk-version }}-${{ matrix.numpy-version }}
```
Or maybe a general for loop to loop over the different matrix options.

Reply:
> Maybe it's possible to simply use `matrix` generically?

So it's possible to use `${{ toJSON(matrix) }}` to get the matrix vars as a string. But then this needs to be parsed to create a valid file name.

E.g. the JSON might be:
``` json
{                                                                                   
  "python-version": "3.8",
  "vtk-version": "9.0.3",
  "numpy-version": "default"
}
```

Surprisingly, github actions don't have any string replacement expressions. Instead, we can use a `bash` function to format this. E.g. define a function `format` that takes `JSON` input, formats it as hyphen-separate name-value pairs, and replaces newline characters with underscores:
``` bash
format() { echo $1  | jq -r 'to_entries[] | "\(.key)-\(.value)"' | tr '\n' '_' | sed 's/_$//' }
```

The problem with this is that I haven't found a way to easily define a bash function  with github actions for re-use by different jobs. Either a separate `.sh` script file needs to be written then called, or encapsulate the formatting as its own separate action that is then called by each job. 

Instead of using a function, we can call this directly, E.g.
``` bash
 echo '{
  "python-version": "3.8",
  "vtk-version": "9.0.3",
  "numpy-version": "default"
}' | jq -r 'to_entries[] | "\(.key)-\(.value)"' | tr '\n' '_' | sed 's/_$//'
```
will output:
```
python-version-3.8_vtk-version-9.0.3_numpy-version-default
```

But then the script needs to be copy/pasted. Not sure if having a more complex bash script copied multiple times is better maintenance-wise than referencing the vars explicitly with
``` bash
${{ github.job }}-${{ matrix.python-version }}-${{ matrix.vtk-version }}-${{ matrix.numpy-version }}
```

Reply:
Right so, the latest commit https://github.com/pyvista/pyvista/pull/6204/commits/e1a54d81743db004533d8dcf602813bd2c81e94b didn't work, the bash needs to be run as its own step to store the filename as a variable first.

This is definitely becoming more complicated than necessary, I think the original suggestions from @tkoyama010 are probably good enough

Reply:
> I found some other questions related to this:
> 
> https://stackoverflow.com/questions/71593861/how-to-get-github-actions-matrix-job-name-within-job

Turns out this answer from the linked question had the winning solution for us
https://stackoverflow.com/a/71971218